stepper-lib/src/air/fold.rs

```diff
@@ -153,17 +120,46 @@ impl<'i> super::ExecutableInstruction<'i> for Next<'i> {
 
         // get the same fold state again because of borrow checker
         match exec_ctx.data_cache.get_mut(iterator_name) {
-            Some(AValue::JValueFoldCursor(fold_state)) => fold_state.cursor -= 1,
+            // move iterator back to provide correct value for possible subtree after next
+            // (for example for cases such as right fold)
+            Some(AValue::JValueFoldCursor(fold_state)) => fold_state.iterable.prev(),
             _ => unreachable!("iterator value shouldn't changed inside fold"),
         };
 
         Ok(())
     }
 }
 
+fn cleanup_variables(exec_ctx: &mut ExecutionCtx<'_>, iterator: &str) {
+    let fold_state = match exec_ctx.data_cache.remove(iterator) {
+        Some(AValue::JValueFoldCursor(fold_state)) => fold_state,
+        _ => unreachable!("fold cursor is changed only inside fold block"),
+    };
+
+    for (variable_name, _) in fold_state.met_variables {
+        exec_ctx.data_cache.remove(variable_name);
+    }
+    exec_ctx.met_folds.pop_back();
+
+    // TODO: fix 3 or more inner folds behaviour
```

Comment:
Can you elaborate in the comment on the current state of things? How will current behavior break on 3 inner folds? How it should be fixed?